### PR TITLE
AO-20081-Refactor-GitHub-Actions-Test-Publish-Workflows-7

### PR DESCRIPTION
**Pull request by ronilan**
This Pull Request is the **7th** in a series of Pull Requests to refactor GitHub Actions (AO-20081). 

It adds a release workflow that is `npm` and `git` triggered, using npm builtin `version` command and `git push tag`. No github hooks needed.

Definitions:
* Build Group are images on which the various versions of the add-on are built. They include combinations to support different Node versions and libc implementations. Generally build is done with the lowest versions of the OS supported, so that `glibc`/`musl` versions are the oldest/most compatible.
* Target Group images include a wide variety of OS and Node version combinations. Group includes both images that can build from code as well as those which can not.

Usage:
* To Release:
  1. On branch run `npm version {major/minor/patch}`(e.g. `npm version patch`) then have the branch pass through the Push/Pull/Merge flow above. 
  2. When ready `git push` origin {tag name} (e.g. `git push origin v11.2.3`).
* Pushing a semantic versioning tag for a patch/minor/major versions (e.g. `v11.2.3`) or an alpha tagged pre-release (e.g. `v11.2.3-alpha.2`) will trigger release.yml. Pushing other pre-release tags (e.g. `v11.2.3-7`) is ignored.

* Workflow will: 
  - Build the code pushed in each of the Build Group images. 
  - Package the built code and upload a tarball to the *production* S3 buckets. 
  - Create all Target Group images and install the prebuilt tarball on each.
  - Publish an NPM package upon successful completion of all steps above. When version tag is `alpha`, package will be NPM tagged same. When it is a release version, package will be NPM tagged `latest`.
* Workflow ensures node-pre-gyp setup is working in *production* for a wide variety of potential customer configurations.
* Workflow publishing to NPM registry exposes the NPM package (and the prebuilt tarballs in the *production* S3 bucket) to the public.
* Note: @appoptics/apm-bindings is not meant to be directly consumed. It is developed as a dependency of [appoptics-apm](https://www.npmjs.com/package/appoptics-apm).

Diagram:

```
push semver tag ─► ┌────────────────────────────┐ ─► ─► ─►
push alpha tag     │Build Group Build & Package │ S3 Package
                   └┬───────────────────────────┘ Production
                    │
                    │   ┌────────────────────┐     │
                    └─► │Target Group Install│ ◄── ▼
                        └┬───────────────────┘
                         │
                         │   ┌───────────┐
                         └─► │NPM Publish│
                             └───────────┘
```

Todos:
  1. Consider adding a dry-run option - will require deleting tarballs from production and running NPM publish with --dry-run. Benefits? Maybe...

Notes:
  * Publish is done from a runner with node 14.x installed. Choice of os/node version has *no* impact (at last ;) )
  * actions/setup-node@v2 handles `.npmrc` file creation for authentication. NPM_AUTH_TOKEN is never written to file.

GitHub Actions bugs:

Pushing a tag triggers not only the Release workflow but also the Build Docker Images workflow even though that one should only be triggered with:
```
on:
  push:
    paths:
      - '.github/docker-node/*.Dockerfile'
```
Adding `tags-ignore:` to the above in attempt to "circumvent" the bug will disable trigger all together. 
Will investigate further (and/or open issue with https://github.com/actions/runner/). 

**Decision was made to disable the trigger on the  Build Docker Images workflow. Images will be built by manual trigger.** (62f51ead1407568ffe385017de621a6c69fe6693)

### CODE REVIEW

File: .github/workflows/docker-node.yml
Comment (kuerbis-martin) [inline .github/workflows/docker-node.yml:4]: Interesting find @ronilan!

File: .github/workflows/release.yml
Comment (kuerbis-martin) [inline .github/workflows/release.yml:11]: I think either a dry-run option or the option to publish to staging instead would certainly be helpful :) although this is tricky with and automated trigger through push. 

I am wondering if you considered a mechanism where you do the release only through a manual workflow trigger, i.e., you can create the tag as part of this workflow, and the user would only provide the desired (version) tag as an input to the workflow? This way, a user would not need to know that pushing a tag will automatically release the artifacts and an unintentional release (by pushing a tag accidentally) would also be prevented.

File: .github/workflows/release.yml
Comment (kuerbis-martin) [inline .github/workflows/release.yml:116]: This won't be needed at the moment, as there is no manual trigger option.

**Review by kuerbis-martin — COMMENTED** [review]
Thanks for the nice write-up @ronilan! I think it would be well worth to add the PR description to some readme to explain the release process, so this information is readily available without searching for the PR which introduced the mechanics. 

I have one (fundamental) comment regarding the general trigger of the release, other than that the changes look good to me! However, please not I am not qualified to judge the actual npm-related code, hence I won't approve this PR and let @cheempz do the final call on this.

File: .github/workflows/docker-node.yml
Comment (ronilan) [reply .github/workflows/docker-node.yml:4]: Issue opened: https://github.com/actions/runner/issues/1202 (and now I'm even more worried as to what triggers when)

File: .github/workflows/release.yml
Comment (ronilan) [reply .github/workflows/release.yml:11]: We can't really "publish to staging" per se. We do staging bucket on each pull request merge. We now publish to production bucket and then, if all is well, do the NPM publish which is technically independent. Versioning imposes a 'one time' success for both bucket and NPM. No overwrite is allowed. This behavior can be overridden but is highly discouraged. Which really leaves doing a dry run with minimal benefits (`--dry-run` flag is not really a full publish run, it stops before the publish step). Which is why it is left in the air...

What we do have is the (still) "secret" dev repo that works with two dev buckets and publishes a dev named NPM package using same code and workflows. There are some gotchas with deleting dev packages, but this is probably the best staging run we can get from such two-stage release process.

So - remove the TODO?

As for triggering via tag - comment below.

File: .github/workflows/release.yml
Comment (ronilan) [reply .github/workflows/release.yml:116]: Correct. Given how trigger filtering work (or don't) I'm a little (irrationally maybe?) hesitant to remove it though...

**Comment by ronilan** [comment]
> Thanks for the nice write-up @ronilan! I think it would be well worth to add the PR description to some readme to explain the release process, so this information is readily available without searching for the PR which introduced the mechanics.

Documentation including all the above is WIP on: https://github.com/appoptics/appoptics-bindings-node/tree/AO-20192-Update-Documentation-no-build

> I have one (fundamental) comment regarding the general trigger of the release

Generally multiple ways to do releases, branches, tags, triggers etc. It's agreed, I think, that there is no right answer on this. The specific method you suggest (which I understand is how the [Python Agent is released](https://github.com/librato/python-appoptics/blob/master/.github/workflows/create_release.yaml)) uses GitHub Release feature. This means workflow manages the versioning and GitHub does `git commit`.

NPM [versioning](https://docs.npmjs.com/cli/v7/commands/npm-version) is good (very, actually) and the way versioning is/was done for this package (and actually any other I've seen before) is using that tool. GitHub Actions does not modify anything and does not commit to the repo. There are (obviously) some edge cases that the developer should be aware of (they'll go in documentation) but this is (generally, I think) the "node way".

File: .github/workflows/release.yml
Comment (kuerbis-martin) [reply .github/workflows/release.yml:11]: Okay thanks for the explanation! From what I understand, it might indeed be the best choice not to implement a dry-run in this case. It might just unnecessarily complicate the workflow with a negligible benefit (if the dry-run code/ mechanics deviated too much from the actual publishing mechanics). I'd probably remove the dry-run TODO altogether then.

Comment (ronilan) [reply .github/workflows/release.yml]: Done.

File: .github/workflows/release.yml
Comment (ronilan) [reply .github/workflows/release.yml]: Left the conditional, changed the comment.

**Review by cheempz — APPROVED** [review]
LGTM, thanks for the great discussions, swooping in here with the easy "hit approve button" part :)

**Pull request merged**
